app/src/main/java/com/example/ncov19traking/ui/graphs/GraphsFragment.kt

```diff
@@ -30,11 +31,14 @@ class GraphsFragment : Fragment() {
     ): View? {
         val root = inflater.inflate(R.layout.fragment_notifications, container, false)
         val chart: LineChart = root.findViewById(R.id.chart)
+        val progressBar: ProgressBar = root.findViewById(R.id.graph_fragment_progressBar)
         val textColor = ContextCompat.getColor(root.context, R.color.graphTextColor)
         setUpChart(chart, textColor)
+        chart.visibility = View.GONE
         graphsViewModel.nCoVAllHistoricalData.observe(viewLifecycleOwner, Observer { nCovTimeline ->
-            val allLineData = ArrayList<ILineDataSet>()
+            progressBar.visibility = View.VISIBLE
             if (nCovTimeline != null) {
+                val allLineData = ArrayList<ILineDataSet>()
                 allLineData.add(
```

Comment:
I know this is not part of your PR, but I think It would be a good Idea to create a method for handling this
```
                allLineData.add(
                    defineDataSet(
                        addDataToEntriesArrays(nCovTimeline.deaths),
                        getString(R.string.total_deaths),
                        ContextCompat.getColor(root.context, R.color.deathsColor)
                    )
                )
```
this piece of code is repeated almost identical 3 times,  I think It will be a good practice for you to abstract this to a method, and reduce the ammount of repeated code.

Reply:
you mean a high order method? because there is actually 3 methods, and were used 3 times because there are 3 dataset

Reply:
that method block is repeated 3 times, in line `43`, `50` and `57`, abstract it, or at least separate the preparation of `allLineData` array to a separate method.

Reply:
the `onCreateView` will become something like this
```
override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_notifications, container, false)
        val chart: LineChart = root.findViewById(R.id.chart)
        val textColor = ContextCompat.getColor(root.context, R.color.graphTextColor)
        setUpChart(chart, textColor)
        graphsViewModel.nCoVAllHistoricalData.observe(viewLifecycleOwner, Observer { nCovTimeline ->
            val allLineData = getLineData(nCovTimeline, root)
            chart.data = LineData(allLineData)
            chart.invalidate()
        })
        graphsViewModel.getErrorOnFetchFailure().observe(viewLifecycleOwner, Observer { error ->
            showErrorMessage(error)
        })
        return root
    }
```
we abstracted all the data logic from that method, the new method `getLineData`

```
private fun getLineData(
        nCovTimeline: Timeline?,
        root: View
    ): ArrayList<ILineDataSet> {
        val allLineData = ArrayList<ILineDataSet>()
        if (nCovTimeline != null) {
            allLineData.addDataSet(nCovTimeline.cases, R.string.global_cases_string, R.color.colorAccent, root.context)
            allLineData.addDataSet(nCovTimeline.deaths, R.string.total_deaths, R.color.deathsColor, root.context)
            allLineData.addDataSet(nCovTimeline.recovered, R.string.total_recovered, R.color.recoveredColor, root.context)
        }
        return allLineData
    }
```
we abstracted even more logic into an extension function called `addDataSet`

```
private fun ArrayList<ILineDataSet>.addDataSet(
        setMap: LinkedHashMap<String, Int>,
        stringAsset: Int,
        colorAsset: Int,
        context: Context,
    ) {
        add(
            defineDataSet(
                addDataToEntriesArrays(setMap),
                getString(stringAsset),
                ContextCompat.getColor(context, colorAsset)
            )
        )
    }
```